### ISSUE TITLE
Questions about the IRLS

### ISSUE THREAD

**Issue by zhanghua7099** [brentyi/jaxls#26]
Hi! 

I found that this project may support IRLS. 

https://github.com/brentyi/jaxls?tab=readme-ov-file#jaxls

Does this mean we can use a robust kernel, such as the Huber and Cauchy function? I searched for this project but couldn't find it. Could you please give me some advice?

Thank you for your help!

**Comment by brentyi** [comment]
Hello!

For IRLS we just need to apply a loss-dependent weight term, right?

You could do this by modifying the `compute_residual=` function that factors take as input. Here's a utility for Huber, I generated from Claude and haven't tested exhaustively but it looks correct to me:

```python
from typing import Callable, Literal, cast

import jax
import jax.numpy as jnp
from jax import numpy as jnp

def huber_loss_wrapper[T: Callable[..., jax.Array]](
    residual_fn: T, k: float = 1.345, mode: Literal["per_term", "norm"] = "per_term"
) -> T:
    """Wraps a residual function with Huber loss for IRLS.

    Args:
        residual_fn: Original residual function
        k: Huber loss parameter (default=1.345 for 95% efficiency on normal distribution)
        mode: Whether to apply weighting per term or based on residual norm

    Returns:
        A new residual function that applies Huber weighting
    """

    def huber_residual(*args):
        residual = residual_fn(*args)
        eps = 1e-8  # epsilon to prevent NaNs

        if mode == "per_term":
            abs_r = jnp.abs(residual) + eps
            w = jax.lax.stop_gradient(jnp.where(abs_r > k, k / abs_r, 1.0))
        else:  # norm
            r_norm = jnp.linalg.norm(residual) + eps
            w = jax.lax.stop_gradient(jnp.where(r_norm > k, k / r_norm, 1.0))
        return residual * jnp.sqrt(w)  # type: ignore

    return cast(T, huber_residual)
```

Usage: `jaxls.Factor(lambda ..., ...)` can be replaced by `jaxls.Factor(huber_loss_wrapper(lambda ...), ...)`.

**Comment by zhanghua7099** [comment]
Hi! Thank you very much for your quick reply. 

I will test this function.

**Comment by zhanghua7099** [comment]
Hi! I‘ve checked this code but I'm still a little confused. 

In my understanding, if we use the coefficient `k=1.345` for the Huber function in IRLS, the residuals $r_i$ should be whitened by current residuals' $(\mu, \sigma)$:

${r}_i'=\dfrac{r_i - \mu}{\sigma}$

and then applying the weighting function. However, this code does not whiten the residuals and directly apply the weighted function. 

Besides, it seems to directly use the current residuals to compute the weight. The weighting value should be determined by the previous iteration's residual. The details can be found in this paper, Sec. 9.4, Eq. (27):

> Zhang, Zhengyou. "Parameter estimation techniques: A tutorial with application to conic fitting." Image and vision Computing 15.1 (1997): 59-76.

Could you please give me some suggestions? 

Thank you very much!

**Comment by brentyi** [comment]
Hi again!

**Whitening:** we don't have this because `jaxls` doesn't implement any whitening. If you have an optimization problem that needs whitening (typically weighting terms, or multiplying with `cov^{-1/2}` when doing MAP estimation with Gaussians) you can handle that yourself in `residual_fn()`, which will be called before the weighting happens.

**Previous iteration's residual:** I only looked briefly, but this looks consistent to me. The "current" residual we compute in `residual = residual_fn(*args)` here is using parameters from the previous iteration. It seems impossible to compute the residual for the current iteration because we don't have the updated parameters yet.
